### PR TITLE
Always enable pthread flags

### LINKED ISSUE #5192: in some cases, GAP must be built with `-pthread` in LDFLAGS
### Observed behaviour

On Fedora 34, 
`Semigroups` package test `SemigroupsTestAll();`
fails with `Error, Enable multithreading to use std::thread: Operation not permitted`

### Expected behaviour

it should pass

-----

This is GAP 4.12.1 built with 
```
$ gcc -v
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-11.3.1-20220421/obj-x86_64-redhat-linux/isl-install --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.3.1 20220421 (Red Hat 11.3.1-2) (GCC) 
```
on x86_64 running Fedora 34, with ld `GNU ld version 2.35.2-6.fc34`. 
Configuring GAP with `./configure LDFLAGS=-pthread <...>` cures the issue.

This appears to be platform-specific, e.g. on current Gentoo (with gcc 12.2.0, ld 2.39.0)  the same test passes.

See https://github.com/semigroups/Semigroups/issues/771#issuecomment-1304459598 for more details.

### PR DESCRIPTION
**Pull request by fingolfin**
Resolves #5192

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1003,9 +1003,15 @@ dnl check for non-standard math functions
 AC_CHECK_FUNCS([exp10])
 
 dnl pthreads
-AS_IF([test "x$enable_hpcgap" = xyes -o "x$with_julia" != xno ],[
+dnl we unconditionally check for these flags (and use them), even though we
+dnl only use pthreads if HPC-GAP is enabled. Als, if a kernel extension uses
+dnl pthreads, then it may be necessary under certain Linux variants that the
```
Comment (dimpase) [inline configure.ac:1008]: Als -> Also

**Review by dimpase — CHANGES_REQUESTED** [review]
modulo the typo, see another comment, all is good

File: configure.ac [hunk 2]
```diff
@@ -1003,9 +1003,15 @@ dnl check for non-standard math functions
 AC_CHECK_FUNCS([exp10])
 
 dnl pthreads
-AS_IF([test "x$enable_hpcgap" = xyes -o "x$with_julia" != xno ],[
+dnl we unconditionally check for these flags (and use them), even though we
+dnl only use pthreads if HPC-GAP is enabled. Als, if a kernel extension uses
```
Comment (fingolfin) [inline configure.ac:1007]:
```suggestion
dnl only use pthreads if HPC-GAP is enabled. Also, if a kernel extension uses
```

**Review by dimpase — APPROVED** [review]

**Comment by fingolfin** [comment]
Hmmm, seems this breaks the `expect` test. Someone else will have to look into this, I don't have the time and interest right now

**Comment by fingolfin** [comment]
[The output](https://github.com/gap-system/gap/actions/runs/6049387897/job/16416579099?pr=5196#step:14:29) is not super helpful:
```
Running test suite testexpect
+ case $TEST_SUITE in
+ INPUTRC=/tmp/inputrc
+ expect -c 'spawn ./gap --quitonbreak -A -b  --cover coverage/testexpect.coverage' /home/runner/work/gap/gap/dev/gaptest.expect
spawn ./gap --quitonbreak -A -b --cover coverage/testexpect.coverage
 TIMEOUT 
Error: Process completed with exit code 2.
```
Of course on my own Linux machines, it just passes with this PR.


I also verified that `-pthread` is used both in the CI tests and on my own machines; I also made sure to pass `--coverage` to compiler and linker alike, just as in the CI environment. We can see the effect in the logs:
```
checking whether pthreads work with "-pthread" and "-lpthread"... yes
checking for joinable pthread attribute... PTHREAD_CREATE_JOINABLE
checking whether more special flags are required for pthreads... no
checking for PTHREAD_PRIO_INHERIT... yes
...
/bin/bash ./libtool --mode=compile --tag CC gcc  -MQ build/obj/src/ariths.c.lo -MMD -MP -MF build/deps/src/ariths.c.d  -pthread --coverage [...]
...
libtool: compile:  gcc -MQ build/obj/src/ariths.c.lo -MMD -MP -MF build/deps/src/ariths.c.d -pthread --coverage [..]
```

Yet despite all this, the expect test just passes on my own machines, while it gets stuck inside CI...

**Comment by dimpase** [comment]
I'm trying to reproduce this locally, using [act](https://github.com/nektos/act). More specifically, I'm using `gh act` installed as a `gh` extension 
```
$ gh extension install https://github.com/nektos/gh-act
```
and running 
```
$ gh act -p false -r -j test
```

`

**Comment by dimpase** [comment]
OK, I can reproduce this in `gh act`. Here I added `-d` to the `expect` options to tell it to output a bit more.
It looks as if GAP never comes to `gap>` prompt.
```diff
| Running test suite testexpect
| + case $TEST_SUITE in
| + INPUTRC=/tmp/inputrc
| + expect -d -c 'spawn ./gap --quitonbreak -A -b  --cover coverage/testexpect.coverage' /home/dimpase/work/software/gap/dev/gaptest.expect
| expect version 5.45.4
| spawn ./gap --quitonbreak -A -b --cover coverage/testexpect.coverage
| parent: waiting for sync byte
| parent: telling child to go ahead
| parent: now unsynchronized from child
| spawn: returns {7327}
| argv[0] = expect  argv[1] = -d  argv[2] = -c  argv[3] = spawn ./gap --quitonbreak -A -b  --cover coverage/testexpect.coverage  argv[4] = /home/dimpase/work/software/gap/dev/gaptest.expect  
| set argc 0
| set argv0 "/home/dimpase/work/software/gap/dev/gaptest.expect"
| set argv ""
| executing commands from command file /home/dimpase/work/software/gap/dev/gaptest.expect
| 
| expect: does "" (spawn_id exp4) match glob pattern "gap> "? no
| expect: timed out
|  TIMEOUT 
[CI/testexpect -  - ubuntu-latest]   ❌  Failure - Main Run tests
[CI/testexpect -  - ubuntu-latest] exitcode '2': failure
```

I've modified the `CI.yml` as follows, to only run `testexpect`.
```diff
--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -44,101 +44,9 @@ jobs:
         shell: [bash]
         test-suites:
           [
-            # base test: fast first test
-            "testinstall",
-
-            "teststandard",
-
-            # run tests contained in the manual
-            "testmanuals",
-
             # test error reporting and compiling as well as libgap
-            "testexpect testmockpkg testspecial test-compile testlibgap testkernel",
-
-            # compile packages and run GAP tests
-            # don't use --enable-debug to prevent the tests from taking too long
-            "testpackages testinstall-loadall",
+            "testexpect",
           ]
-        extra: [""]
-
-        # add a few extra tests
-        include:
-          # test creating the manual
-          # also test `make install` here, as that wants the manual, too;
-          # and set NO_COVERAGE=1 to ensure the installed GAP binaries don't
-          # contain references to the build directory
-          - os: ubuntu-latest
-            shell: bash
-            test-suites: "makemanuals testmakeinstall"
-            extra: "GAPPREFIX=/tmp/gapprefix CONFIGFLAGS=\"--prefix=/tmp/gapprefix\" NO_COVERAGE=1"
-
-          - os: ubuntu-20.04
-            shell: bash
-            test-suites: "teststandard"
-            extra: "ABI=32 CONFIGFLAGS=\"\""
-
-          # FIXME: we used to run `teststandard` for HPC-GAP under Travis CI,
-          # but somehow when running on GitHub Actions, it takes almost 4
-          # hours (!) to complete instead of 25 minutes. So for now we just
-          # run testinstall.
-          - os: ubuntu-latest
-            shell: bash
-            test-suites: "testinstall"
-            extra: "HPCGAP=yes ABI=64"
-
-          # this job also tests GAP without readline
-          - os: macos-latest
-            shell: bash
-            test-suites: "testmockpkg testinstall"
-            extra: "BOOTSTRAP_MINIMAL=yes"
-
-          # run bugfix regression tests
-          # Also turn on '--enable-memory-checking' to make sure GAP compiles
-          # with the flag enabled. We do not actually test the memory
-          # checking, as this slows down GAP too much.
-          # Also check ubuntu-latest works.
-          - os: ubuntu-latest
-            shell: bash
-            test-suites: "testbugfix"
-            extra: "CONFIGFLAGS=\"--enable-memory-checking\""
-
-          # out of tree builds -- these are mainly done to verify that the
-          # build system work in this scenario. Since we don't expect the test
-          # results to vary compared to the in-tree builds, we turn off
-          # coverage reporting by setting NO_COVERAGE=1; this has the extra
-          # benefit of also running the tests at least once with the
-          # ReproducibleBehaviour option turned off.
-
-          # The '--enable-valgrind' checks that GAP builds and runs correctly
-          # when compiled with valgrind support. We do not actually run any
-          # tests using valgrind, as it is too slow.
-          - os: ubuntu-latest
-            shell: bash
-            test-suites: "testbuildsys testmockpkg testinstall"
-            extra: "NO_COVERAGE=1 ABI=64 BUILDDIR=out-of-tree
-                    CONFIGFLAGS=\"--enable-valgrind\""
-            packages: "valgrind"
-
-          # same as above, but in 32 bit mode, also turn off debugging (see
-          # elsewhere in this file for an explanation).
-          - os: ubuntu-20.04
-            shell: bash
-            test-suites: "testbuildsys testmockpkg testinstall"
-            extra: "NO_COVERAGE=1 ABI=32 BUILDDIR=out-of-tree CONFIGFLAGS=\"\""
-
-          # test Julia integration
-          - os: ubuntu-20.04
-            shell: bash
-            test-suites: "testinstall"
-            extra: "JULIA=yes CONFIGFLAGS=\"--enable-debug --disable-Werror\""
-
-          - os: windows-2019
-            # The 'run' steps in this job use Cygwin's bash, once it is set up.
-            #   --login: make a login shell (so PATH is set up)
-            #   -o igncr: Accept windows line endings
-            #   {0} : Pass any extra arguments from CI
-            shell: C:\cygwin64\bin\bash.exe --login -o igncr '{0}'
-            test-suites: "testmockpkg testinstall"
 
     env:
       TEST_SUITES: ${{ matrix.test-suites }}
```

**Comment by ChrisJefferson** [comment]
I believe this is fixed now. I found this fix completely by accident while fiddling around using `gh act` (which is a nice thing to learn about). Turns out turning on pthreads is somehow making github CI take a REALLY long time to start GAP, so we need a huge timeout.

**Review by dimpase — APPROVED** [review]
LGTM, also passes in my `gh act`.

**Comment by james-d-mitchell** [comment]
Thanks everyone, much appreciated!

**Comment by fingolfin** [comment]
While the `testexpect` issue is "resolved" (yay!) this still can't be merged as now three other CI jobs fail due to timeouts: whereas they normally take 25-30 minutes, they now did not finish after 1 hour. Something fishy is going on here.

I'll force another CI run to see if this is persistent (even if all tests passt, please do not merge blindly but rather let's verify CI job times have not increased a lot)

**Comment by dimpase** [comment]
the order of the libraries linked/loaded might play a role. We recently saw weirdness related to this in https://github.com/sagemath/sage/issues/31572

**Comment by dimpase** [comment]
it can be specific to glibc. Do you see such a slowdown in macOS instead of ubuntu-latest? or on Void linux (which has musl libc) ?

**Pull request closed without merging**